peps/pep-0788.rst

```diff
@@ -0,0 +1,619 @@
+seems to fit well for the functions in this PEP.
+
+Full deprecation of old APIs
+----------------------------
+
+As made clear in Motivation_, ``PyGILState`` is already pretty buggy, and
+even if it was magically fixed, the current behavior of hanging the thread is
+beyond repair. As such, this PEP intends to completely deprecate the existing
+``PyGILState`` APIs. However, even if this PEP is rejected, all of the APIs
+can be replaced with more correct ``PyThreadState`` functions in the current
+C API:
+
+- :c:func:`PyGILState_Ensure`: :c:func:`PyThreadState_Swap` / :c:func:`PyThreadState_New`
+- :c:func:`PyGILState_Release`: :c:func:`PyThreadState_Clear` / :c:func:`PyThreadState_Delete`
+- :c:func:`PyGILState_GetThisThreadState`: :c:func:`PyThreadState_Get`
+- :c:func:`PyGILState_Check`: ``PyThreadState_GetUnchecked() != NULL``
+
+A light layer of magic
+----------------------
+
+The APIs proposed by this PEP intentionally have a layer of "magic" that is
+kept from the user and offloads complexity onto CPython maintainers. This is
+done primarily to help ease the transition from ``PyGILState`` for existing
+codebases, and for ease-of-use to those who provide wrappers the C API, such
+as Cython or PyO3. See also :ref:`pep-788-activate-deactivate-instead`.
+
+Specification
+=============
+
+Interpreter reference counting
+------------------------------
+
+Internally, the interpreter will have to keep track of a reference count
+field, which will determine when the interpreter state is actually
+deallocated. This is done to prevent use-after-free crashes in
+:c:func:`PyThreadState_Ensure` for interpreters with short lifetimes.
+
+An interpreter state returned by :c:func:`Py_NewInterpreter` (or really,
+:c:func:`PyInterpreterState_New`) will start with a reference count of 1, and
+:c:func:`PyInterpreterState_Delete` will decrement the reference count. If the
+new reference count is zero, :c:func:`PyInterpreterState_Delete` will
+deallocate the interpreter state. However, the reference count will *not*
+prevent the interpreter from finalizing.
+
+.. c:function:: PyInterpreterState *PyInterpreterState_Hold(void)
+
+    Similar to :c:func:`PyInterpreterState_Get`, but returns a strong
+    reference to the interpreter (meaning, it has its reference count
+    incremented by one, allowing the returned interpreter state to be safely
+    accessed by another thread).
+
+    This function is generally meant to be used in tandem with
+    :c:func:`PyThreadState_Ensure`.
+
+    The caller must have an :term:`attached thread state`, and cannot return a
+    failure.
+
+
+.. c:function:: void PyInterpreterState_Release(PyInterpreterState *interp)
+
+    Decrement the reference count of the interpreter. This function mainly
+    exists for completeness, and should rarely be used; nearly all references
+    returned by :c:func:`PyInterpreterState_Hold` should be released by
+    :c:func:`PyThreadState_Ensure`.
```

Comment:
I don't think that it's the second sentence is useful:

```suggestion
    Decrement the reference count of the interpreter.
```

Or you may document that this function can be called when PyInterpreterState_Hold() was called but PyThreadState_Ensure() cannot be called. I don't know how to explain that simply :-)